rxandroidble/src/main/java/com/polidea/rxandroidble/internal/connection/RxBleGattCallback.java

```diff
@@ -51,35 +52,33 @@ public Object call(BleGattException bleGattException) {
         }
     };
-    private final Observable disconnectedErrorObservable = gattAndConnectionStateOutput.valueRelay
-            .filter(new Func1<Pair<BluetoothGatt, RxBleConnectionState>, Boolean>() {
-                @Override
-                public Boolean call(Pair<BluetoothGatt, RxBleConnectionState> pair) {
-                    return isDisconnectedOrDisconnecting(pair);
-                }
-            })
-            .map(new Func1<Pair<BluetoothGatt, RxBleConnectionState>, Object>() {
-                @Override
-                public Object call(Pair<BluetoothGatt, RxBleConnectionState> bluetoothGattRxBleConnectionStatePair) {
-                    throw new BleDisconnectedException(bluetoothGattRxBleConnectionStatePair.first.getDevice().getAddress());
-                }
-            })
-            .mergeWith(gattAndConnectionStateOutput.errorRelay.map(errorMapper))
-            .replay()
-            .autoConnect(0);
+    private final Observable disconnectedErrorObservable;
 
     @Inject
     public RxBleGattCallback(@Named(ClientComponent.NamedSchedulers.BLUETOOTH_CALLBACKS) Scheduler callbackScheduler,
                              BluetoothGattProvider bluetoothGattProvider,
-                             NativeCallbackDispatcher nativeCallbackDispatcher) {
+                             NativeCallbackDispatcher nativeCallbackDispatcher,
+                             BluetoothDevice bluetoothDevice) {
         this.callbackScheduler = callbackScheduler;
         this.bluetoothGattProvider = bluetoothGattProvider;
         this.nativeCallbackDispatcher = nativeCallbackDispatcher;
-    }
 
-    private boolean isDisconnectedOrDisconnecting(Pair<BluetoothGatt, RxBleConnectionState> pair) {
-        RxBleConnectionState rxBleConnectionState = pair.second;
-        return rxBleConnectionState == RxBleConnectionState.DISCONNECTED || rxBleConnectionState == RxBleConnectionState.DISCONNECTING;
+        final Observable<?> isDisconnectedOrDisconnecting = gattAndConnectionStateOutput.valueRelay.filter(
+                new Func1<Pair<BluetoothGatt, RxBleConnectionState>, Boolean>() {
+                    @Override
+                    public Boolean call(Pair<BluetoothGatt, RxBleConnectionState> pair) {
+                        RxBleConnectionState rxBleConnectionState = pair.second;
+                        return rxBleConnectionState == RxBleConnectionState.DISCONNECTED
+                                || rxBleConnectionState == RxBleConnectionState.DISCONNECTING;
+                    }
+                }
+        );
+        final Observable<?> otherConnectionErrors = gattAndConnectionStateOutput.errorRelay.map(errorMapper);
+        this.disconnectedErrorObservable = Observable.error(new BleDisconnectedException(bluetoothDevice.getAddress()))
+                .delaySubscription(isDisconnectedOrDisconnecting)
+                .mergeWith(otherConnectionErrors)
+                .replay()
+                .autoConnect(0);
```

Comment:
As a matter of fact, did we analyze if this observable is unsubscribed at some point of time? `autoConnect` is a risky operator.

Reply:
This should not be a problem as every connection related object is being garbage collected once the connection is no longer used